### PR TITLE
UCS/DEBUG: Log zero-size allocation.

### PR DESCRIPTION
**Pull request by rakhmets**
## What
Added logging of zero-size allocation.

### CODE REVIEW

File: src/ucs/debug/memtrack.c [hunk 1]
```diff
@@ -27,6 +27,12 @@
 #define UCS_MEMTRACK_FORMAT_STRING    ("%22s: size: %9lu / %9lu\tcount: %9u / %9u\n")
 
 
+#define UCS_MEMTRACK_LOG_ZERO_SIZE_ALLOACTION(_size, _name) \
+    if (_size == 0) { \
```
Comment (yosefe) [inline src/ucs/debug/memtrack.c:31]: use () for macro params

File: src/ucs/debug/memtrack.c [hunk 2]
```diff
@@ -27,6 +27,12 @@
 #define UCS_MEMTRACK_FORMAT_STRING    ("%22s: size: %9lu / %9lu\tcount: %9u / %9u\n")
 
 
+#define UCS_MEMTRACK_LOG_ZERO_SIZE_ALLOACTION(_size, _name) \
```
Comment (yosefe) [inline src/ucs/debug/memtrack.c:30]: maybe also log the pointer?

File: src/ucs/debug/memtrack.c [hunk 3]
```diff
@@ -27,6 +27,12 @@
...
+    if (_size == 0) { \
+        ucs_debug("zero-size allocation %s", _name); \
```
Comment (yosefe) [inline src/ucs/debug/memtrack.c:32]: smth like "allocated zero-size block %p for %s"

File: test/gtest/ucs/test_memtrack.cc [hunk 4]
```diff
@@ -263,3 +263,56 @@ UCS_TEST_F(test_memtrack, custom) {
...
+
+    static size_t log_count;
+    static ucs_log_level_t orig_log_level;
+};
+
+size_t test_memtrack_log::log_count = 0;
+ucs_log_level_t test_memtrack_log::orig_log_level;
+
+UCS_TEST_F(test_memtrack_log, zero_size) {
+    orig_log_level = ucs_global_opts.log_component.log_level;
```
Comment (yosefe) [inline test/gtest/ucs/test_memtrack.cc:294]: maybe add option to set/restore log level to scoped_log_handler, as optional second parameter to the constructor? for example default param value is UCS_LOG_LEVEL_LAST which means do-nothing.

**Review by yosefe — COMMENTED** [review]
how does it help to have this debug message?

File: test/gtest/ucs/test_memtrack.cc (hunk 4 above)
Comment (rakhmets) [reply test/gtest/ucs/test_memtrack.cc]: Replaced `ucs_debug` by `ucs_warn`.

File: test/gtest/ucs/test_memtrack.cc [hunk 5]
```diff
@@ -263,3 +263,46 @@ UCS_TEST_F(test_memtrack, custom) {
...
+    static ucs_log_func_rc_t
+    zero_size_log_handler(const char *file, unsigned line, const char *function,
+                          ucs_log_level_t level,
+                          const ucs_log_component_config_t *comp_conf,
+                          const char *message, va_list ap)
+    {
+        std::string log_str = format_message(message, ap);
+        if ((level == UCS_LOG_LEVEL_WARN) &&
+            (log_str.find("allocated zero-size block") != std::string::npos)) {
+            ++log_count;
```
Comment (yosefe) [inline test/gtest/ucs/test_memtrack.cc:279]: maybe print the message using UCS_TEST_MESSAGE?
it's done in wrap_errors_logger for example

**Review by yosefe — DISMISSED** [review]

File: src/ucs/debug/memtrack.c [hunk 6]
```diff
@@ -311,13 +317,15 @@ static void ucs_memtrack_vfs_init()
...
     void *ptr = malloc(size);
+    UCS_MEMTRACK_LOG_ZERO_SIZE_ALLOACTION(size, ptr, name);
     ucs_memtrack_allocated_internal(ptr, size, name);
     return ptr;
 }
 
 void *ucs_calloc(size_t nmemb, size_t size, const char *name)
 {
     void *ptr = calloc(nmemb, size);
+    UCS_MEMTRACK_LOG_ZERO_SIZE_ALLOACTION(size, ptr, name);
```
Comment (brminich) [inline src/ucs/debug/memtrack.c:328]: looks like we need to pass `nmemb*size`

**Review by yosefe — DISMISSED** [review]

**Comment by yosefe** [comment]
@rakhmets can you pls squash?

**Comment by rakhmets** [comment]
> @rakhmets can you pls squash?

@yosefe I'm waiting for https://github.com/openucx/ucx/pull/8152. https://github.com/openucx/ucx/pull/8152 contains fix for zero-size allocation of ucp_dst_md_cmpts. After that it is ok to squash and merge the pull request.

**Review by yosefe — APPROVED** [review]

**Pull request merged**